### ISSUE TITLE
[Performance]: Server launched with LoRA adapter but queried without is much slower than standard server

### ISSUE THREAD

**Issue by matkle** [vllm-project/vllm#10898]
### Report of performance regression

There are several existing issues related to the performance of LoRA, but I haven't seen the problem that I encountered. 

When launching a server without any lora adapter via 
`vllm serve meta-llama/Llama-2-7b-hf --tensor_parallel_size 8 --max-model-len 4096`
and querying it via 
`time curl http://localhost:8000/v1/completions \
      -H "Content-Type: application/json" \
      -d '{
          "model": "meta-llama/Llama-2-7b-hf",
          "prompt": "Repeat the following sentence two hundred times: San Francisco is a city in California.",
          "max_tokens": 3000,
          "temperature": 0
      }'`,
I observe a real time of about **15 seconds** (on a machine with eight A100s with 40GB).

But when I launch a server with lora adapter via
`vllm serve meta-llama/Llama-2-7b-hf --tensor_parallel_size 8 --max-model-len 4096 --enable-lora --lora-modules lora=$HOME/.cache/huggingface/hub/models--yard1--llama-2-7b-sql-lora-test/... --fully-sharded-loras` 
and query _the base model only_ (i.e., without the lora adapter) in the very same way as above, I measure a real time of about **23 seconds**.

I would like to understand where this additional running time is coming from? I observe significant additional running time whether I set `--enforce-eager` or specify additional arguments like `--max-loras`, `--max-lora-rank`, etc. or not.

**Comment by jeejeelee** [comment]
Have you tried removing `--fully-sharded-loras`?

**Comment by matkle** [comment]
Yes, then I measure **18 seconds**. So much less overhead, but still some overhead. I would like to understand where this overhead is coming from (and for me the fully sharded version is more relevant since ultimately I want to use larger models and larger adapters).

**Comment by jeejeelee** [comment]
If you set `--enable-lora`, even when you serve  the `base_model`, it will still involve calls to LoRA-related ops, which will incur overhead and result in approximately 10% performance loss

**Comment by matkle** [comment]
And why is that? Is that really necessary at query time?

**Comment by jeejeelee** [comment]
> And why is that? Is that really necessary at query time?

Yes. Considering we need to use cudagraph to reduce the overhead of the LoRA kernel.

**Comment by paolovic** [comment]
The same for me, it's much more a 10x than a 10% performance loss for me :(

**Comment by jeejeelee** [comment]
> The same for me, it's much more a 10x than a 10% performance loss for me :(

could you plz provibe your running script?

**Comment by paolovic** [comment]
Hi @jeejeelee ,
thank you for your fast response.
I will do asap and also do some tests to prove my point with numbers

**Comment by Nero10578** [comment]
> Hi @jeejeelee , thank you for your fast response. I will do asap and also do some tests to prove my point with numbers

Make sure you disable eager mode and set max seq capture length to the max seq length so that you are using cuda graphs.

**Comment by Nero10578** [comment]
> Have you tried removing `--fully-sharded-loras`?

So is using fully-sharded-loras supposed to make it faster or slower? The docs seem to imply it is faster, but in my testing it is slower.
